src/mscorlib/shared/System/Memory.cs

```diff
@@ -0,0 +1,313 @@
+                return new Memory<T>(Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory), _index + start, _length - start);
+            return new Memory<T>(Unsafe.As<T[]>(_arrayOrOwnedMemory), _index + start, _length - start);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given memory, beginning at 'start', of given length
+        /// </summary>
+        /// <param name="start">The index at which to begin this slice.</param>
+        /// <param name="length">The desired length for the slice (exclusive).</param>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> or end index is not in range (&lt;0 or &gt;=Length).
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Memory<T> Slice(int start, int length)
+        {
+            if ((uint)start > (uint)_length || (uint)length > (uint)(_length - start))
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+
+            // There is no need to 'and' _index by the bit mask here 
+            // since the constructor will set the highest order bit again anyway
+            if (_index < 0)
+                return new Memory<T>(Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory), _index + start, length);
+            return new Memory<T>(Unsafe.As<T[]>(_arrayOrOwnedMemory), _index + start, length);
+        }
+
+        /// <summary>
+        /// Returns a span from the memory.
+        /// </summary>
+        public Span<T> Span
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                if (_index < 0)
+                    return Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory).AsSpan(_index & bitMask, _length);
+                return new Span<T>(Unsafe.As<T[]>(_arrayOrOwnedMemory), _index, _length);
+            }
+        }
+
+        public unsafe MemoryHandle Retain(bool pin = false)
+        {
+            MemoryHandle memoryHandle;
+            if (pin)
+            {
+                if (_index < 0)
+                {
+                    memoryHandle = Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory).Pin();
+                }
+                else
+                {
+                    var handle = GCHandle.Alloc(Unsafe.As<T[]>(_arrayOrOwnedMemory), GCHandleType.Pinned);
+                    void* pointer = Unsafe.Add<T>((void*)handle.AddrOfPinnedObject(), _index);
+                    memoryHandle = new MemoryHandle(null, pointer, handle);
+                }
+            }
+            else
+            {
+                if (_index < 0)
+                {
+                    Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory).Retain();
+                    memoryHandle = new MemoryHandle(Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory));
+                }
+                else
+                {
+                    memoryHandle = new MemoryHandle(null);
```

Comment:
If Retain is meant to keep the object alive, then it seems like we still need to store the array into the handle, so as to keep a reference to it.  Imagine, for example, the array contained finalizable objects; if I did:
```C#
T[] array = ...;
MemoryHandle h = new Memory<T>(array).Retain();
array = null;
...
```
I would expect as long as I'm still holding onto `h` that the objects in the `array` wouldn't get finalized, but with this code as it stands, they could be finalized the moment I set `array = null`.

Reply:
Will this work?
```C#
var array = Unsafe.As<T[]>(_arrayOrOwnedMemory);
void* pointer = Unsafe.Add<T>((void*)Unsafe.AsPointer(ref array[0]), _index);
memoryHandle = new MemoryHandle(null, pointer);
```

Reply:
> Will this work?

No.  Nothing is keeping the array alive here.

Reply:
As implemented today (and envisioned), Retain is not a guarantee that the buffer will be kept alive. It just guarantees that the reference the caller has to the buffer will work (as opposed to being disposed/deleted/etc).

I am not saying that this is the right behavior we should ship, but rather that this is what Retain was originally designed to do.

Reply:
If we wanted to change the behavior/design, we could change the MemoryHandle._owner field to be type of Object and store either IRetainable or T[] in the filed. It would keep the array alive. But it could keep the array alive for longer than needed in some cases. It's a tradeoff.

Reply:
I don't have a strong opinion, either.  Just seemed strange that Retain() doesn't actually "retain" in that case.

Reply:
As to pinning, when the pin parameter is true, we create GC handle for the array, and so I am not sure I understand the comments about the pointer not being pinned. It's pinned when pin == true and it's not pinned when it's false (but then the pointer property returns null)

Reply:
> so I am not sure I understand the comments about the pointer not being pinned

The comments are in response to Ahson's "will this work" example.